### ISSUE TITLE
Add new values to a bayesian network.

### ISSUE THREAD

**Issue by gnzlo789** [jmschrei/pomegranate#47]
Hello,

I am starting to learn bayesian networks, and this seems a great project to work with. This is my code:

```python
activity = DiscreteDistribution({
	'breakfast': 1./6,
	'lunch': 1./6,
	'dinner': 1./6,
	'hangout': 1./6,
	'call': 1./6,
	'meeting': 1./6
})

duration = ConditionalProbabilityTable([
	['breakfast', 30, 1.0],
	['breakfast', 60, 0.0],
	['lunch', 30, 0.5],
	['lunch', 60, 0.5],
	['dinner', 30, 0.0],
	['dinner', 60, 1.0],
	['hangout', 30, 0.0],
	['hangout', 60, 1.0],
], [activity])

# Make the states
s0 = State( activity, name="type")
s1 = State( duration, name="duration")

# Make the bayes net, add the states, and the conditional dependencies.
network = BayesianNetwork( "example" )
network.add_states( [s0, s1] )
network.add_transition( s0, s1 )
network.bake()
```

My question is how can I add a new activity to the Discrete Distribution, and to the ConditionalProbabilityTable?

Thanks.

**Comment by jmschrei** [comment]
Howdy

The easiest way would be to define a new instance. There's no current way to add new keys to existing objects which is intuitive. Is this something which wouldn't work in your case?

**Comment by gnzlo789** [comment]
I could do it that way, but I will have to redefine the duration table (and others that I omitted). And then call the train method (I will have hundreds/thousands of training data for each user (1 user has 1 BN)), and with every action the user take, I will have to update the BN to learn from the users preference. I think creating everything again could lead to a really big overhead in a multi-user environment.

I also have another ConditionalProbabilityTable that has hundreds/thousands of rows. Do you have any suggestion on how to manage them?

The problem is that I have to define all the cases from the beginning (the probability of many of them is 0), and there is no way to update it (add new keys) without recreating it. For example I want to add a new preference:
```python
duration.train([['meeting',30,0.2]]) # this will do nothing, 'meeting' is not in the definition of duration #1 post
duration.add_train([['meeting',30,0.2]]) # something like this is it possible?
```
And 'add_train' it will add/update only the rows related. Otherwise I have to send all the data I have of the user on every update. 

Is it possible to add this functionality? Or I am the only one that has this problem.

Sorry for my English. Thanks in advance.

**Comment by jmschrei** [comment]
Let me just confirm your problem. You want to add new keys to a distribution which is a parent distribution to another node, and so you'd have to update both the parent and child distribution. But adding keys is dynamic, and so you can't define it at the beginning? Would it be possible to just add the full set of keys at the beginning, just with probability 0? 

It may be possible for me to add this functionality, but not soon, as I'm swamped with other forms of work recently. Would it be possible for you to include a minimal code snippet showing me exactly what you'd like to see added?

**Comment by gnzlo789** [comment]
Hi!

>You want to add new keys to a distribution which is a parent distribution to another node, and so you'd have to update both the parent and child distribution. 

Correct.

>But adding keys is dynamic, and so you can't define it at the beginning? Would it be possible to just add the full set of keys at the beginning, just with probability 0?

No, because I don't know the information.

I created a new method called add_items (Distribution, DiscreteDistribution, ConditionalProbabilityTable). It adds new items with a probability 0. There is an example in the code below.
You can see what I added here: https://github.com/gnzlo789/pomegranate/blob/master/pomegranate/distributions.pyx

I think it works well for DiscreteDistribution, but I got an error with ConditionalProbabilityTable. The problem occurs when executes line 88 (code below). 
The error is because it is trying to do: d[key] = value / total. And total is 0.
I have been thinking of two ways of solving this:
1. If an item doesn't exist then the result should be 0. (d[key]=0).
2. I have to change the add_items method to create every possibility.

I haven't try to solve that problem yet. Maybe tomorrow I will try option 1, which seems to me the right one.

>Would it be possible for you to include a minimal code snippet showing me exactly what you'd like to see added?

```python
from pomegranate import *

# Meeting Type
meeting_type = DiscreteDistribution({
	'breakfast': 1./2,
	'lunch': 1./2,
	# 'dinner': 1./6,
	# 'hangout': 1./6,
	# 'call': 1./6,
	# 'meeting': 1./6
})

duration = ConditionalProbabilityTable([
	['breakfast', 30, 1.0],
	['breakfast', 60, 0.0],
	['lunch', 30, 0.5],
	['lunch', 60, 0.5],
	# ['dinner', 30, 0.0],
	# ['dinner', 60, 1.0],
	# ['hangout', 30, 0.0],
	# ['hangout', 60, 1.0],
], [meeting_type])


# Make the states
s0 = State( meeting_type, name="type")
s1 = State( duration, name="duration")

# Make the bayes net, add the states, and the conditional dependencies.
network = BayesianNetwork( "meeting" )
network.add_states( [s0, s1] )
network.add_transition( s0, s1 )
network.bake()


# The first observation is that the duration is thirty minutes
first_observation = { 'duration' : 30 }

# beliefs will be an array of posterior distributions or clamped values for each state, indexed corresponding to the order
# in self.states.
beliefs = network.forward_backward( first_observation )

# Convert the beliefs into a more readable format
beliefs = map( str, beliefs )

# Print out the state name and belief for each state on individual lines
print "\n".join( "{}\t{}".format( state.name, belief ) for state, belief in zip( network.states, beliefs ) )


### Testing the new add_items method

print "== Meeting Type BEFORE =="
print meeting_type

print "== Add new key to meeting type =="
meeting_type.add_items( ['call'] )
meeting_type.train( ['call'], inertia=0.7 )

print "== Meeting Type AFTER =="
print meeting_type

print

print "== Duration BEFORE =="
print duration

print "== Add new row to the table =="
duration.add_items( [['call', 30]] )
duration.add_items( [['call', 120]] )
duration.train( [['call', 120], ['call', 30]] )

print "== Duration AFTER =="
print duration

print

print "== Network Training =="
# data = [[ 'call', 30, 1.0 ],]
data = [[ 'call', 30, 1.0 ],]
network.train( data )
print "== Meeting Type =="
print meeting_type
print "== Duration =="
print duration

# The second observation is that the meeting type is a 'meeting'
# second_observation = { 'type' : 'call' } # Works
second_observation = { 'duration' : 120 } # ERROR!

# beliefs will be an array of posterior distributions or clamped values for each state, indexed corresponding to the order
# in self.states.
beliefs = network.forward_backward( second_observation )

# Convert the beliefs into a more readable format
beliefs = map( str, beliefs )

# Print out the state name and belief for each state on individual lines
print "\n".join( "{}\t{}".format( state.name, belief ) for state, belief in zip( network.states, beliefs ) )
```

**Comment by gnzlo789** [comment]
Hello,

I added the functionality I needed. Now I can add new items to distributions, I also added a new method in the BayesianNetwork class, it's called update. When invoked, it checks that exists every possible combination of values given the parents distributions. So there is not no need to define every possibility at the beginning.

Of course after doing this it's necessary to call the train method.

Bye.
